### ISSUE TITLE
Question about SwiGLU Implementation: Differences between Transformer-based Encoder code and paper description

### ISSUE THREAD

**Issue by hanseul-jeong** [ai4co/routefinder#4]
Thank you for your excellent work!

Could you please point me to where the SwiGLU part is implemented in the Transformer-based Encoder? 
I believe it's in the ParallelGatedMLP class, but it seems to be implemented differently from what's described in the paper's appendix. Specifically, I noticed differences regarding the presence of bias terms, the use of sigmoid function, and the final multiplication with x.
I would appreciate any clarification or explanation about these implementation differences and the reasoning behind them.

Thanks in advance for your help! :)

**Comment by hanseul-jeong** [comment]
"I'm focusing on Cross Problem Learning for Combinatorial Optimization, and I'm working on enhancing your model :)
Thank you for your prompt response!
I see your reviewer is quite detail-oriented with your work 😄
I have an additional question regarding the attention function. In your implementation, I noticed that logit_k is calculated and then multiplied to produce the final logit.
Is this approach documented in your paper? I'm curious if this is a specific technique you've found to improve performance?
Thanks again for your help!"

**Comment by fedebotu** [comment]
> I have an additional question regarding the attention function. In your implementation, I noticed that logit_k is calculated and then multiplied to produce the final logit.
Is this approach documented in your paper? 

Yes! These are at page 19, eq. 18

> I'm curious if this is a specific technique you've found to improve performance?

This is actually called the Pointer Mechanism, which refers back to [Pointer Networks](https://arxiv.org/abs/1506.03134)!

**Comment by hanseul-jeong** [comment]
Hello, I think I may have phrased my question confusingly. 
My question was specifically about the cached.logit_key that is calculated from the encoder output. 
Looking at the PointerAttention class, there's this line:

`logits = (torch.bmm(glimpse, logit_key.squeeze(-2).transpose(-2, -1))).squeeze(-2) / math.sqrt(glimpse.size(-1))`

I understood that z in equation 18 corresponds to the glimpse, but I was curious about this additional multiplication with logit_key. If I've misunderstood something, I'd appreciate clarification.

Additionally, while it might be out of place to ask here, I'm curious about how the embeddings in Appendix D were calculated. 
It seems like you plotted one embedding for each of the 100 instances - are these the averages of all node embeddings, or did you use only the depot node embeddings?

Thank you!

**Comment by fedebotu** [comment]
> I understood that z in equation 18 corresponds to the glimpse, but I was curious about this additional multiplication with logit_key. If I've misunderstood something, I'd appreciate clarification.


Yes, that's correct! This refers to the paper from [Bello et al. 2016](https://arxiv.org/abs/1611.09940). Note that this mechanism is also used in the seminal paper from Kool et al. 2019 (here in the [code](https://github.com/wouterkool/attention-learn-to-route/blob/c9abf41ac2f878a55b20dc7e829bc942bb999631/nets/attention_model.py#L477))

> Additionally, while it might be out of place to ask here, I'm curious about how the embeddings in Appendix D were calculated.
It seems like you plotted one embedding for each of the 100 instances - are these the averages of all node embeddings, or did you use only the depot node embeddings?

You may now check the code for these here [here](https://github.com/ai4co/routefinder/blob/main/examples/3-tsne.ipynb)~
